### PR TITLE
[DependencyInjection] Deprecate scope concept

### PR DESCRIPTION
**Pull request by dosten**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| License | MIT |

This PR mark as deprecated the concept of scopes in the DI container. See https://github.com/symfony/symfony/issues/10557#issuecomment-38811755.
Also adds a new `shared` flag to the service definitions in replacement of the `prototype` scope.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 1]
```diff
@@ -191,6 +191,10 @@ public function setParameter($name, $value)
      */
     public function set($id, $service, $scope = self::SCOPE_CONTAINER)
     {
+        if (self::SCOPE_CONTAINER !== $scope) {
+            @trigger_error('The concept of container scopes is deprecated since version 2.8 and will be removed in 3.0. Use only the "container" scope.', E_USER_DEPRECATED);
```
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: Any valid reason for a leading `@` sign?

File: src/Symfony/Component/DependencyInjection/Definition.php [hunk 2]
```diff
@@ -608,6 +609,10 @@ public function getFile()
      */
     public function setScope($scope)
     {
+        if (ContainerInterface::SCOPE_CONTAINER !== $scope) {
+            @trigger_error('The concept of container scopes is deprecated since version 2.8 and will be removed in 3.0. Use only the "container" scope.', E_USER_DEPRECATED);
```
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Definition.php]: Same as above.

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Container.php]: see #14900

Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: Thanks, I didn't see this PR ;)

**Comment by stof** [comment]
You need to make sure that core code does not trigger such deprecation though (for the ``request`` service for instance).

and the loaders should trigger their own deprecations when using ``scope`` to have a better message (same than done in a few other places)

**Comment by dosten** [comment]
@stof The only cases in the core code are the `request` scope and the `test.client` service with the `prototype` scope. What's the best approach?

**Comment by fabpot** [comment]
@dosten As you can see on Travis, there are still many tests that need to be moved as legacy ones. For the `test.client`, I think that we need to keep the `prototype` scope as it was done before the introduction of scopes. Back then, we had a flag for prototype. As we decided to call it `factory` in Pimple (as this is where prototypes make more sense), I think we can add a `factory` flag which would replace the `prototype` scope.

**Comment by dosten** [comment]
@fabpot Thank you for your feedback, I didn't have enough time to continue with this PR. I'll work on this soon. 
I've searched for the flag that you refer, and I have found this commit 1d5b6ed908046d2765b585c05b7ab4697a8c2b3d. Are you refering to the `shared` flag?
IMO if we add a `factory` flag this would be confusing, we already have the concept of factories in the DI. `shared` is a good name IMO.

**Comment by dosten** [comment]
I've added a initial implementation of a `shared` flag in the service definition, so any service that sets this flag as `false` will return a new instance every time, like the `prototype` scope. There is many failing tests because the deprecation messages, but i want to discuss this before fix the tests.

**Comment by dosten** [comment]
I can't trigger an error on the methods of the `ContainerInterface` because the BC promise, any idea?

**Comment by fabpot** [comment]
@dosten I was indeed refering to the old `shared` flag (and you are right that using factory like in Pimple is probably confusing). ping @symfony/deciders

**Comment by Tobion** [comment]
`shared` makes sense to me.

**Comment by dosten** [comment]
@fabpot This PR is ready to be reviewed.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 3]
```diff
@@ -191,6 +193,10 @@ public function setParameter($name, $value)
      */
     public function set($id, $service, $scope = self::SCOPE_CONTAINER)
     {
+        if (!in_array($scope, array('container', 'request')) || ('request' === $scope && 'request' !== $id)) {
```
Comment (dosten) [inline src/Symfony/Component/DependencyInjection/Container.php]: I'm not sure about this condition, maybe something like `'request' !== $id` is enough.

**Comment by Tobion** [comment]
:+1:  good job

File: src/Symfony/Component/DependencyInjection/Definition.php [hunk 4]
```diff
@@ -598,6 +605,34 @@ public function getFile()
...
+     *
+     * @param bool $shared Whether the service must be shared or not
+     *
+     * @return Definition The current instance
+     *
+     * @api
+     */
+    public function setShared($shared)
+    {
+        $this->shared = (bool) $shared;
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Definition.php]: Do we really want to cast here? The docblock is explicit about the expected type.

File: src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php [hunk 5]
```diff
@@ -68,9 +68,9 @@ public function getProxyFactoryCode(Definition $definition, $id)
     {
         $instantiation = 'return';
 
-        if (ContainerInterface::SCOPE_CONTAINER === $definition->getScope()) {
+        if ($definition->isShared() && ContainerInterface::SCOPE_CONTAINER === $definition->getScope(false)) {
```
Comment (xabbuh) [inline src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php]: We will need to bump the DependencyInjection component version in `composer.json` of the bridge or check if the method exists first.

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 -----
 
  * allowed specifying a directory to recursively load all configuration files it contains
+ * deprecated the concept of scopes
+ * added support of non shared service definitions
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: non-shared

File: src/Symfony/Component/DependencyInjection/Definition.php (hunk 4 above)
Comment (dosten) [inline src/Symfony/Component/DependencyInjection/Definition.php]: We actually cast in the other methods of the `Definition` class

File: src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php (hunk 5 above)
Comment (dosten) [inline src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php]: I prefer to check if the method exists.

File: UPGRADE-2.8.md [hunk 7]
```diff
@@ -65,3 +65,48 @@ Translator
...
+   After:
+
+   ```php
+   use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+   $container = new ContainerBuilder();
+   $container
+       ->register('foo', 'stdClass')
+       ->setShared(false)
+   ;
```
Comment (fabpot) [inline UPGRADE-2.8.md]: I think we need to show examples for YML and XML as well as those formats are the most used ones for configuring services.

Comment (dosten) [inline UPGRADE-2.8.md]: Agree, the structure should be?

Before:

- PHP example
- Yaml example
- XML example

After

- PHP example
- Yaml example
- XML example

?

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 -----
 
  * allowed specifying a directory to recursively load all configuration files it contains
+ * deprecated the concept of scopes
+ * added support of non-shared service definitions
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: can you mention the new `shared` attritube?

**Comment by fabpot** [comment]
:+1:

File: UPGRADE-2.8.md (hunk 7 above)
Comment (fabpot) [inline UPGRADE-2.8.md]: Looks good to me

**Comment by dosten** [comment]
@fabpot Comments addressed

**Comment by fabpot** [comment]
Thank you @dosten.

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 9]
```diff
@@ -163,8 +163,15 @@ private function parseDefinition($id, $service, $file)
...
         }
 
+        if (isset($service['shared'])) {
+            $definition->setShared($service['shared']);
+        }
+
         if (isset($service['scope'])) {
-            $definition->setScope($service['scope']);
+            if ('request' !== $id) {
+                @trigger_error(sprintf('The "scope" key in file "%s" is deprecated since version 2.8 and will be removed in 3.0.', $file), E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: the deprecation message should mention the service id IMO to make it easier to understand

Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: see #15113

**Pull request merged**